### PR TITLE
Allow sharex/y after axes creation.

### PR DESCRIPTION
**Pull request by anntzer**
This intentionally does not allow unsharing or changing the shared axes,
as there are bigger questions on the API there (#9923 and issues linked there).

The API is named `Axes.sharex()` to allow for a later `Axes.unsharex()`,
though (`set_sharex()` would be fine, but `set_unsharex()`? or perhaps
`set_sharex(None)`, though).  Happy to hear suggestions.

An example use case is creating a grid with `subplots()`, but with
custom sharing relationships between the subplots -- e.g., sharex/sharey
across all, except the first row of axes which only shares x with their
column and the first column which only shares y with their lines.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/axes/_base.py [hunk 1]
```diff
@@ -945,6 +946,42 @@ def _gen_axes_spines(self, locations=None, offset=0.0, units='inches'):
         return OrderedDict((side, mspines.Spine.linear_spine(self, side))
                            for side in ['left', 'right', 'bottom', 'top'])
 
+    def sharex(self, other):
+        """
+        Share the x-axis with *other*.
+
+        This is equivalent to passing ``sharex=other`` when constructing the
```
Comment (jklymak) [inline lib/matplotlib/axes/_base.py]: This docstring could be more explicit and non circular.  For instance the verb to share would indicate that self shares with other, whereas the exact opposite happens.  Further passing sharex=other calls this function so I don’t think it’s very helpful to say that this function is the same as that and expect the user to ferret out the behaviour

**Comment by anntzer** [comment]
I'm explicitly *not* allowing toggling, or changing who you share with, just going from "not-shared" to "shared".  Indeed, allowing unsharing makes the problem much more complex.

The use case hinted above is:
```
  A A A
B X X X
B X X X
B X X X
```
Each A plots the distribution (histogram) of variable *a* in conditions a1, a2, a3; each B plots the distribution of variable *b* in conditions b1, b2, b3, and each X plots the joint distribution (scatter plot).  Here you want to sharex across rows and sharey across columns, *except* that the first row does not share y's and the first column does not share x's (let's say you don't want to normalize your histograms).
Right now if you want to do this you basically need to give up on subplots() and manually add axes in a loop while keeping track of who to share with what -- and that info needs to be passed immediately.
With the PR here, you can start by calling subplots(), then create the required share-linkages.

**Comment by jklymak** [comment]
I understand your example, but it would be helpful if you actually provided one in the PR.

**Comment by anntzer** [comment]
Do you mean one as comments in the code?  (where?)  in the commit message?

**Comment by jklymak** [comment]
I meant as an example in the example folder

File: lib/matplotlib/axes/_base.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/axes/_base.py]: "self shares with other, whereas the exact opposite happens"
What do you mean by "opposite"?  Isn't sharing symmetrical? (well perhaps not deep in the implementation details, but from the user PoV it is)

**Comment by anntzer** [comment]
Sure, added a simple example.

File: lib/matplotlib/axes/_base.py (hunk 1 above)
Comment (jklymak) [reply lib/matplotlib/axes/_base.py]: I meant in everyday English the transitive verb “to share” means one person shares Something with another

Comment (anntzer) [reply lib/matplotlib/axes/_base.py]: Can you propose an alternate wording here?

Comment (jklymak) [reply lib/matplotlib/axes/_base.py]: Sorry, the main point is that saying `ax.sharex` is "Equivalent to passing sharex=other when constructing..." is a circular doc, because when you construct, you call `ax.sharex`.  This function ideally would define what `ax.sharex` does explicitly.

Comment (anntzer) [reply lib/matplotlib/axes/_base.py]: Right now the docs of sharex (as kwargs to the Axes constructor) are
```
        sharex, sharey : `~.axes.Axes`, optional
            The x or y `~.matplotlib.axis` is shared with the x or
            y axis in the input `~.axes.Axes`.
```
which is not much better :/ I agree this can be improved, but is arguably a separate issue?

Comment (jklymak) [reply lib/matplotlib/axes/_base.py]: The docstring defines what the function does for the user, and one of our main problems with "share" is that people don't know what it means and even in the history of the code base different definitions have popped up.  Right now I'm not 100% sure what this PR changed and I think it should be defined clearly, in plain English, what calling this function will do.

Comment (anntzer) [reply lib/matplotlib/axes/_base.py]:
```
ax0 = (some axes)
ax1 = fig.add_subplot(...)
ax1.sharex(ax0)
```
is equivalent to
```
ax0 = (some axes)
ax1 = fig.add_subplot(..., sharex=ax0)
```
Now it is true that what the latter does is not really well documented either, but the doc available for `ax1.sharex(ax0)` is, well, exactly the same.

**Comment by efiring** [comment]
First thoughts on this:
1) The API, with a simple `sharex()` method, is reasonable.  I would not advocate prepending `set_`.  More descriptive alternatives would be `sharex_with()` and `sharex_from()`.  The latter has the advantage of noting that info (objects to be shared) will come from the argument, not go to it.
2) It seems like unsharing could be added by supporting the `None` argument, and replacing the shared objects with deep copies of them; but I haven't looked at this in detail.
3) Leaving aside the unsharing, which I agree does not have to be addressed here, I think the present approach would work fine when used early--immediately after Axes creation--but inevitably users will find ways to make it surprise them.  For example, if two Axes are created, each is called with `set_aspect(1, adjustable='datalim')`, and then some time later they are shared, the problem won't be caught until `draw()`, when `apply_aspect()` is called.

**Comment by anntzer** [comment]
I guess I'm +0.5 on sharex_with() and +0.25 on sharex_from() (both are better than sharex()), let's see what others think.

What about adding something like

> Note that this method is only intended to be called immediately after the axes has been created; unexpected behavior may occur if the axes already has its own aspect ratio, formatters, or locators set up.

to the docstrings?  (I don't really intend to support "later" calls with this PR anyways.)

**Comment by anntzer** [comment]
rebased

**Comment by efiring** [comment]
We seem to be getting a systematic failure--not just on this PR--in the macosx build on Travis, in `test_bbox_inches_tight_raster[pdf]`, an image comparison failure.
In addition there are the two circleci docs failures here.  Any idea what is going on? @QuLogic?

**Comment by QuLogic** [comment]
macOS failure is #17065; doc failure is just spurious inability to download a Sphinx inventory.
